### DISCUSSION TITLE
failed to install brownie

### DISCUSSION THREAD

**Discussion by Jimmy321342** [smartcontractkit/full-blockchain-solidity-course-py#370] [Q&A] [answered]
pip failed to build package:
    cytoolz

Some possibly relevant errors from pip install:
    build\lib.win-amd64-3.10\cytoolz\functoolz.cp310-win_amd64.pyd : fatal error LNK1120: 1 unresolved externals
    error: command 'C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\BuildTools\\VC\\Tools\\MSVC\\14.29.30133\\bin\\HostX86\\x64\\link.exe' failed with exit code 1120

Error installing eth-brownie.

**Comment by cromewar** [comment, marked as answer]
As I mentioned in the other thread you have to possible paths to solve this:

1. Install visual studio community and the python development tools on it.
2. install WSL and work with a Linux terminal.

**Reply by sl00k** [reply]
I think I have the same issue :(

![image](https://user-images.githubusercontent.com/33859801/142048973-ce616d39-deb5-4fb1-a030-7aaedf8739dc.png)

![image](https://user-images.githubusercontent.com/33859801/142049103-3674d23b-9afe-46aa-a5a5-461b74f8ea47.png)
second entry should be the native python development tools

**Reply by Jimmy321342** [reply]
PS C:\Users\A\web3_py_simple_storage> pipx install eth-brownie
'eth-brownie' already seems to be installed. Not modifying existing installation in 'C:\\Users\\A\\.local\\pipx\\venvs\\eth-brownie'. Pass '--force' to force installation.
PS C:\Users\A\web3_py_simple_storage> brownie
brownie : The term 'brownie' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, 
verify that the path is correct and try again.
At line:1 char:1
+ brownie
+ ~~~~~~~
    + CategoryInfo          : ObjectNotFound: (brownie:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
 

:(

**Reply by cromewar** [reply]
@Jimmy321342  did you install visual studio and the python development tools?

Also the problem yo have is because windows does not recognize packages installed with pipx, so you must add them to windows environment variables or install using pip.

Note: if you want to get rid off this headaches, use WSL for blockchain (and python) development.

**Reply by Jimmy321342** [reply]
yes I did, i installed visual studio community 2022 and visual studio build tools 2019, installed the python development tools in it, now installing native python in the optional as suggested by @sl00k

**Reply by cromewar** [reply]
try it and see if it works.

**Reply by Jimmy321342** [reply]
PS C:\Users\A\web3_py_simple_storage> pipx install eth-brownie
>>
Pass '--force' to force installation.
PS C:\Users\A\web3_py_simple_storage> pipx upgrade eth-brownie
>>
Not upgrading eth-brownie. It has missing internal pipx metadata.
It was likely installed using a pipx version before 0.15.0.0.
Please uninstall and install this package to fix.


got this

**Reply by cromewar** [reply]
as I said, preferably don't use pipx to install brownie, you'll other problems with vscode later, uninstall pipx and install eth-brownie using just pip.

**Reply by Jimmy321342** [reply]
WARNING: The candidate selected for download or install is a yanked version: 'protobuf' candidate (version 3.18.0 at https://files.pythonhosted.org/packages/74/4e/9f3cb458266ef5cdeaa1e72a90b9eda100e3d1803cbd7ec02f0846da83c3/protobuf-3.18.0-py2.py3-none-any.whl#sha256=615099e52e9fbc9fde00177267a94ca820ecf4e80093e390753568b7d8cb3c1a (from https://pypi.org/simple/protobuf/))
Reason for being yanked: This version claims to support Python 2 but does not
Using legacy 'setup.py install' for cytoolz, since package 'wheel' is not installed.
Installing collected packages: cytoolz, six, eth-utils, varint, urllib3, toml, rlp, pyrsistent, pycryptodome, py, pluggy, parsimonious, packaging, netaddr, multidict, iniconfig, idna, hexbytes, eth-keys, charset-normalizer, certifi, base58, attrs, atomicwrites, yarl, python-dateutil, pytest, multiaddr, jsonschema, inflection, eth-rlp, eth-keyfile, eth-abi, chardet, bitarray, async-timeout, websockets, wcwidth, tomli, sortedcontainers, regex, pywin32, pytest-forked, pyjwt, pygments, protobuf, platformdirs, mythx-models, lru-dict, ipfshttpclient, execnet, eth-account, dataclassy, click, asttokens, aiohttp, wrapt, web3, vyper, vvm, tqdm, pyyaml, pythx, python-dotenv, pytest-xdist, pygments-lexer-solidity, py-solc-x, py-solc-ast, psutil, prompt-toolkit, lazy-object-proxy, hypothesis, eth-event, eip712, black, eth-brownie
  Attempting uninstall: cytoolz
    Found existing installation: cytoolz 0.11.2
    Uninstalling cytoolz-0.11.2:

didn't work with pip, i guess i'll have to look for wsl

**Reply by cromewar** [reply]
It seems it  is trying to install using python version 2, maybe you could use:

`python3 -m pip install eth-brownie`.

Or. better solution use WSL.

**Reply by sl00k** [reply]
and than? what does it change? if we use WSL, do we still code in vsc? is there a different way how we have to install packages?

**Reply by cromewar** [reply]
@sl00k  Yes you'll keep using vscode.
WSL is a whole different operating system running inside you windows machine, the package administration is totally different, please go to the course index and refer the apart "windows support" to know how to setup your windows machine to work with WSL.

**Reply by Jimmy321342** [reply]
python3 -m pip install eth-brownie
python3 : The term 'python3' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, 
verify that the path is correct and try again.
At line:1 char:1
+ python3 -m pip install eth-brownie
+ ~~~~~~~
    + CategoryInfo          : ObjectNotFound: (python3:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

**Reply by Jimmy321342** [reply]
@cromewar tried with wsl, met with following limitations  
![three](https://user-images.githubusercontent.com/92639455/142155667-42ff647f-3102-4a9a-8202-d67924b0836e.PNG)
![one](https://user-images.githubusercontent.com/92639455/142155697-e731ee96-5679-43af-b2df-3c68d4ea7a3b.PNG)


while doin pip install eth-brownie

![two](https://user-images.githubusercontent.com/92639455/142155712-a6535674-b5ab-4f5d-8225-d4a66f635f09.PNG)

**Reply by cromewar** [reply]
@Jimmy321342 

First update your WSL. (update and upgrade commands)

Second.
![image](https://user-images.githubusercontent.com/20199180/142200800-e4941293-1635-452d-89b3-9851f72625a5.png)

The Path on WSL and Windows are completely different and separate, if you want to add something yo for path permanently on linux you should modify the `~/.bashrc` file. so pasthe the same command there.

Thirdly. don't worry about npm audit, is normal

and FInally, don't worry for the packages pip are not upgrading either and remember to use a virtualenv for each project.

**Reply by Jimmy321342** [reply]
'   Requirement already satisfied: pythx==1.6.1 in /home/jim/.local/lib/python3.8/site-packages (from eth-brownie) (1.6.1) '

looks like it is done now :) 
how do I migrate to VScode or i just code in ubuntu ( in which case where's my terminal? coding space? ) excuse me for so many queries I am from a non IT background 

ps i have already installed wsl extension on vscode

**Reply by sl00k** [reply]
here you go https://www.youtube.com/watch?v=CTlTBQU_lro

**Reply by cromewar** [reply]
You just need to click the button left corner and open a WSL window, or from your ubuntu terminal just type `code .`

**Reply by Jimmy321342** [reply]
everything's in place now, a million thanks to @cromewar  and @sl00k , without you people helping me I would have never been able to go forth with this tutorial.

**Reply by cromewar** [reply]
I'm glad this helped, feel free to ask any more questions and also thanks @sl00k for the help you are awesome!
I'll close this thread now.
